### PR TITLE
docs: Improved API documentation for BasicCrawler class

### PR DESCRIPTION
**Pull request by belloibrahv**
### **Description**

This PR improves the API documentation for the `BasicCrawler` class by providing clear and concise explanations for all arguments and methods. The documentation now adheres to Google style guidelines, making it easier for users to understand and use the class effectively.

### **Issues**

- Relates: #304

### **Testing**

Manual testing was conducted to ensure that the API documentation is accurate and informative. The changes do not affect the functionality of the `BasicCrawler` class.

### **Checklist**

- [ ] CI passed

### CODE REVIEW

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 1]
```diff
@@ -336,12 +347,14 @@ async def run(
         *,
         purge_request_queue: bool = True,
     ) -> FinalStatistics:
-        """Run the crawler until all requests are processed.
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:339]: Please stay with the imperative form (`run`, not `runs`).

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 2]
```diff
@@ -336,12 +347,14 @@ async def run(
...
 
         Args:
-            requests: The requests to be enqueued before the crawler starts
-            purge_request_queue: If this is `True` and the crawler is not being run for the first time, the default
-                request queue will be purged
+            requests: A list of requests to enqueue before the crawl starts.
+            purge_request_queue: Whether to clear the request queue before starting the crawler run.
+
+        Returns:
+            FinalStatistics: Contains the final statistics about the crawler run.
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:357]: Do not include types in the docstrings (`FinalStatistics:`), as we have type annotations for that.

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 3]
```diff
@@ -88,14 +88,30 @@ class BasicCrawlerOptions(TypedDict, Generic[TCrawlingContext]):
...
+        use_session_pool (bool, optional): Use session pool (default: True).
+        session_pool (Optional[SessionPool]): Preconfigured session pool.
+        retry_on_blocked (bool, optional): Retry on blocked requests (default: True).
+        proxy_configuration (Optional[ProxyConfiguration]): Proxy configuration.
+        statistics (Optional[Statistics]): Statistics object.
+        event_manager (Optional[EventManager]): Event manager.
+        configure_logging (bool, optional): Configure logging (default: True).
+        _context_pipeline (Optional[ContextPipeline[TCrawlingContext]]): Internal context pipeline.
+        _additional_context_managers (Optional[Sequence[AsyncContextManager]]): Additional context managers.
+        _logger (Optional[logging.Logger]): Logger instance.
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:114]: Please remove types from docstrings. They are redundant, as we have type annotations.

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 4]
```diff
@@ -88,14 +88,30 @@ class BasicCrawlerOptions(TypedDict, Generic[TCrawlingContext]):
...
-
-    The URLs to crawl are fed either from a static list of URLs or from a dynamic queue of URLs enabling recursive
-    crawling of websites.
-
-    `BasicCrawler` is a low-level tool that requires the user to implement the page download and data extraction
-    functionality themselves. If we want a crawler that already facilitates this functionality, we should consider using
-    one of its subclasses.
+    """A versatile web crawler for parallel URL fetching.
+
+    Provides a basic framework for crawling websites, allowing customization of request handling and data extraction.
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:93]: The class docstring should be much longer in this case. Checkout the JS version for inspiration [BasicCrawler]( https://crawlee.dev/api/basic-crawler/class/BasicCrawler). You should describe the features that are provided by the `BasicCrawler` - scaling, request routing, sessions, proxies, retries, statistics ...

**Review by vdusek — CHANGES_REQUESTED** [review]
To be honest, I am not sure whether the `BasicCrawler` is a good starting point, as it is pretty complex. You will be required to go through the code and understand the features it provides. Maybe you can start with some more simple public components.

**Comment by belloibrahv** [comment]
> To be honest, I am not sure whether the `BasicCrawler` is a good starting point, as it is pretty complex. You will be required to go through the code and understand the features it provides. Maybe you can start with some more simple public components.

Hi @vdusek,

Thank you for your feedback on my PR. I appreciate your suggestion to start with a simpler public component.

I've considered your recommendation and believe it would be beneficial to gain a better understanding of the codebase before diving into the BasicCrawler class.

Could you please suggest a suitable component that would be a good starting point for me? I'm open to exploring different options and am eager to contribute to the project in a meaningful way.

Thank you for your guidance!

**Comment by belloibrahv** [comment]
Hi @vdusek,

I've addressed the feedback from your previous review and updated the PR with the following changes:

* Removed type annotations from docstrings.
* Updated the class docstring to provide a more comprehensive description of the `BasicCrawler` class.
* Considered starting with a simpler component as suggested and am open to recommendations.

Please review the changes and let me know if there's anything else I need to address. Thank you for your time and feedback!

**Comment by belloibrahv** [comment]
Hello @netmilk @janbuchar @B4nan  @vdusek, could you kindly review the PR.

File: src/crawlee/basic_crawler/_basic_crawler.py (hunk 1 above)
Comment (vdusek) [reply src/crawlee/basic_crawler/_basic_crawler.py:339]: This applies to all doc titles.

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 5]
```diff
@@ -88,14 +88,56 @@ class BasicCrawlerOptions(TypedDict, Generic[TCrawlingContext]):
...
+    )
+    await crawler.run()
+    ```
+
+    Advanced features:
+    - Configure concurrency to control parallel processing
+    - Set up automatic retry mechanisms for failed or blocked requests
+    - Utilize session pools for efficient connection management
+    - Integrate proxy services for IP rotation
+    - Monitor detailed statistics about the crawling process
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:119]: Could you use sentences (with dots at the end)? Thanks.

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 6]
```diff
@@ -88,14 +88,56 @@ class BasicCrawlerOptions(TypedDict, Generic[TCrawlingContext]):
...
+        use_session_pool: Whether to use session pool. Defaults to True.
+        session_pool: Preconfigured session pool.
+        retry_on_blocked: Whether to retry on blocked requests. Defaults to True.
+        proxy_configuration: Proxy configuration.
+        statistics: Statistics object.
+        event_manager: Event manager.
+        configure_logging: Whether to configure logging. Defaults to True.
+        _context_pipeline: Internal context pipeline.
+        _additional_context_managers: Additional context managers.
+        _logger: Logger instance.
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:140]: Please remove `Defaults to...` as this is contained already in the function declaration. Again, this applies to all docstrings.

**Review by vdusek — CHANGES_REQUESTED** [review]

**Comment by belloibrahv** [comment]
Hello @netmilk @janbuchar @B4nan @vdusek, could you kindly review the PR.

**Review by vdusek — COMMENTED** [review]
Could you please rebase on current master? So we can execute fixed CI here.

**Review by vdusek — CHANGES_REQUESTED** [review]
Also, as I said in the issue, this should be only about class docstrings, not about rewording the method docstrings (although you can add e.g. missing docstring for arg/return if makes sense).

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 7]
```diff
@@ -124,31 +166,25 @@ def __init__(
...
+            use_session_pool: Use session pool.
+            session_pool: Preconfigured session pool.
+            retry_on_blocked: Retry on blocked requests.
+            proxy_configuration: Proxy configuration.
+            statistics: Statistics object.
+            event_manager: Event manager.
+            configure_logging: Configure logging.
+            _context_pipeline: Internal context pipeline.
+            _additional_context_managers: Additional context managers.
+            _logger: Logger instance.
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:187]: What's the purpose of this?

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 8]
```diff
@@ -88,14 +88,56 @@ class BasicCrawlerOptions(TypedDict, Generic[TCrawlingContext]):
...
+        use_session_pool: Whether to use session pool.
+        session_pool: Preconfigured session pool.
+        retry_on_blocked: Whether to retry on blocked requests.
+        proxy_configuration: Proxy configuration.
+        statistics: Statistics object.
+        event_manager: Event manager.
+        configure_logging: Whether to configure logging.
+        _context_pipeline: Internal context pipeline.
+        _additional_context_managers: Additional context managers.
+        _logger: Logger instance.
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:140]: The same info we have in the init. Keep it only in one place.

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 9]
```diff
@@ -88,14 +88,56 @@ class BasicCrawlerOptions(TypedDict, Generic[TCrawlingContext]):
...
+    )
+    await crawler.run()
+    ```
+
+    Advanced features:
+    - Configure concurrency to control parallel processing.
+    - Set up automatic retry mechanisms for failed or blocked requests.
+    - Utilize session pools for efficient connection management.
+    - Integrate proxy services for IP rotation.
+    - Monitor detailed statistics about the crawling process.
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:119]: Please only use sections from Google docs style, see https://github.com/apify/crawlee-python/issues/324 for more information.

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 10]
```diff
@@ -88,14 +88,56 @@ class BasicCrawlerOptions(TypedDict, Generic[TCrawlingContext]):
...
+    Basic usage:
+    ```python
+    async def handle_request(context: Context) -> None:
+        url = context.request.url
+        response = await context.send_request()
+        # Process the response here
+    crawler = BasicCrawler(
+        request_handler=handle_request,
+    )
+    await crawler.run()
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:111]: This is not a recommended way of attaching a handler - typically we use decorators.

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 11]
```diff
@@ -88,14 +88,56 @@ class BasicCrawlerOptions(TypedDict, Generic[TCrawlingContext]):
...
+    BasicCrawler is a highly customizable and efficient web crawling solution that provides:
+    - Automatic scaling based on available system resources and target website's requirements.
+    - Smart request routing and handling of different types of URLs.
+    - Session management for maintaining persistent connections.
+    - Proxy integration and rotation for avoiding IP-based blocking.
+    - Automatic retries for failed requests with customizable settings.
+    - Built-in statistics tracking for monitoring crawler performance.
+    - Event system for fine-grained control over the crawling process.
+
+    Basic usage:
```
Comment (vdusek) [inline src/crawlee/basic_crawler/_basic_crawler.py:102]: This should be called Example.

**Review by vdusek — CHANGES_REQUESTED** [review]

File: src/crawlee/basic_crawler/_basic_crawler.py (hunk 8 above)
Comment (belloibrahv) [reply src/crawlee/basic_crawler/_basic_crawler.py:140]: Acknowledge

**Comment by belloibrahv** [comment]
@vdusek I've updated the BasicCrawler documentation based on your feedback:
- Renamed "Basic usage" to "Example"
- Removed duplicate parameter docs
- Switched to decorator pattern in examples
- Followed Google style guide formatting
Please take a look when you have a moment.

**Comment by vdusek** [comment]
Have you checked the JS API doc of this class? Have you tried to execute the example? It's completely wrong. Sorry, this is not going to work.

**Pull request closed without merging**